mbed_build/_internal/config/cumulative_data.py

```diff
@@ -0,0 +1,66 @@
+#
+# Copyright (C) 2020 Arm Mbed. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Ability to parse cumulative attributes from Sources."""
+import itertools
+import re
+from dataclasses import dataclass, field, fields
+from typing import Any, Iterable, Set, Tuple
+
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from mbed_build._internal.config.source import Source
+
+
+@dataclass
+class CumulativeData:
+    """Representation of cumulative attributes assembled during Source parsing."""
+
+    macros: Set[str] = field(default_factory=set)
+    features: Set[str] = field(default_factory=set)
+    components: Set[str] = field(default_factory=set)
+    labels: Set[str] = field(default_factory=set)
+    device_has: Set[str] = field(default_factory=set)
+
+    @classmethod
+    def from_sources(cls, sources: Iterable["Source"]) -> "CumulativeData":
+        """Interrogate each Source in turn to create final CumulativeData."""
+        data = CumulativeData()
+        for source in sources:
+            for key, value in source.cumulative_overrides.items():
+                _modify_field(data, key, value)
+        return data
+
+
+def _modify_field(data: CumulativeData, key: str, value: Any) -> None:
+    """Mutates CumulativeData in place by adding, removing or resetting the value of a field."""
+    key, modifier = _extract_target_modifier_data(key)
+    if modifier == "add":
+        new_value = getattr(data, key) | set(value)
```

Comment:
Huh, I never new that was used to represent union for sets in python. Thats neat.

Reply:
On the similar subject `operator` is another module full of neat stuff.
```
{'A'} | {'B'} == operator.or_({'A'}, {'B'})
```
Useful if you want a more functional approach to your code.

Example:
```
modifications = {
  'add': operator.or_,
  'remove': operator.sub,
}
modifications['remove']({'A', 'B'}, {'B'})
```